File: src/factory.ml
(* 
 * Copyright (C) 2015 Gaëtan Dubreil
 *
 *  All rights reserved.This file is distributed under the terms of the
 *  GNU General Public License version 3.0.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *)

open Graffophone_plugin
open Usual
open SampleFormat

module Tkr = Talker

let talkerHandlers : (string * Plugin.talkerHandler) list ref = ref []

let addTalkerHandler th = Plugin.(
    talkerHandlers := (th.kind, th) :: !talkerHandlers;
    traceGreen("Talker "^th.kind^" ("^th.category^") registered")
  )

let getTalkerHandler kind =
  try L.assoc kind !talkerHandlers
  with Not_found -> (trace(kind^"'s factory not found!"); raise Not_found)


let makeTalker ?name kind =
  let th = getTalkerHandler kind in

  let talker = Tkr.mkTkr(Plugin.(th.make ())) in

  match name with None -> talker
                | Some nm -> talker#setName nm; talker


let getTalkersInfos() =
  L.map !talkerHandlers ~f:(fun (kind, th) -> Plugin.(kind, th.category))


let outputMakers:(string * Output.handler)list ref = ref []

let addOutputMaker handler =
  outputMakers := Output.(handler.feature, handler) :: !outputMakers

let getOutputMaker feature = Output.((L.assoc feature !outputMakers).make)

let makeOutput name feature attributs =
  (getOutputMaker feature) name attributs

File: src/factory.rs
use std::rc::Rc;

use gpplugin::audio_format::AudioFormat;
use gpplugin::talker::RTalker;

use crate::output::ROutput;
use crate::playback_output::Playback;
use crate::plugins_manager::PluginsManager;

pub struct Factory {
    plugins_manager: PluginsManager,
}

pub type RFactory = Rc<Factory>;

impl Factory {
    pub fn new() -> Factory {
        let mut plugins_manager = PluginsManager::new();
        plugins_manager.load_plugins();
        Self { plugins_manager }
    }

    pub fn new_ref() -> RFactory {
        Rc::new(Factory::new())
    }

    pub fn make_talker(&self, model: &str, name: Option<&str>) -> Result<RTalker, failure::Error> {
        self.plugins_manager.make_talker(model, name)
    }

    pub fn make_output(
        &self,
        _name: &str,
        _kind: &str,
        _attributs: &Vec<(&str, &str, &str)>,
    ) -> Result<ROutput, failure::Error> {
        Playback::new_ref(2, AudioFormat::chunk_size())
    }
}
